### PR TITLE
Refactor inbound_firewall to not depend on smartstack dependencies

### PR DESCRIPTION
**Pull request by y-trobinso**
Previously, this was placed at the end of `_smartstack_rules`, placing inbound filters after rules established to allow smartstack traffic. When envoy is used, this doesn't make sense. It also doesn't make sense to require a dependencies file for this filtering behavior, either.

As such, this has been moved upwards one layer and refactored accordingly. I've also included an allow for traffic from localhost, should other rules not propagate it to this location.

Including Krall as CC, due to some chatter about healthcheck traffic today. This is an opt-in feature, so this should only impact services that include it.

### CODE REVIEW

File: paasta_tools/firewall.py [hunk 1]
```diff
@@ -182,19 +188,29 @@ def _yocalhost_rule(port, comment, protocol="tcp"):
...
+        port = conf.get_proxy_port(service_name, instance_name)
+        yield iptables.Rule(
+            protocol=protocol,
+            src="0.0.0.0/0.0.0.0",
+            dst="0.0.0.0/0.0.0.0",
+            target="REJECT",
+            matches=((protocol, (("dport", (str(port),)),)),),
+            target_parameters=((("reject-with", ("icmp-port-unreachable",))),),
+        )
+        yield iptables.Rule(
```
Comment (danielpops) [inline paasta_tools/firewall.py:206]: do you need moar local IPs like 169.254.255.254 169.254.0.1 etc?

Comment (y-trobinso) [reply paasta_tools/firewall.py:206]: That depends on whether we want to continue to support smartstack traffic (re, our conversation about using this with envoy). If so, we'll also want those two IPs.

I guess they can't hurt regardless. I'll add them.

File: paasta_tools/firewall.py [hunk 2]
```diff
@@ -77,13 +81,19 @@ def get_rules(self, soa_dir, synapse_service_dir):
             # for several minutes after the directory disappears from soa-configs.
             return ()
 
-        if not conf.get_outbound_firewall():
+        if not conf.get_inbound_firewall() and not conf.get_outbound_firewall():
             return ()
```
Comment (kawaiwanyelp) [inline paasta_tools/firewall.py:85]: nit: Are these lines necessary anymore? You're already checking for `get_inbound_firewall` and `get_outbound_firewall` below.

File: paasta_tools/firewall.py [hunk 3]
```diff
@@ -182,19 +192,30 @@ def _yocalhost_rule(port, comment, protocol="tcp"):
...
+    If this is set to "reject", this is limited only to traffic from localhost"""
+    policy = conf.get_inbound_firewall()
+    if policy == "reject":
+        port = conf.get_proxy_port(service_name, instance_name)
+        yield iptables.Rule(
+            protocol=protocol,
+            src="0.0.0.0/0.0.0.0",
+            dst="0.0.0.0/0.0.0.0",
+            target="REJECT",
+            matches=((protocol, (("dport", (str(port),)),)),),
```
Comment (kawaiwanyelp) [inline paasta_tools/firewall.py:207]: Not a review comment. Just an "oof"; why do these parentheses and commas remind me of Lisp...

File: paasta_tools/firewall.py (hunk 2 above)
Comment (y-trobinso) [reply paasta_tools/firewall.py:85]: Since `tuple(list())` should indeed be equivalent, removed. :)

**Comment by y-trobinso** [comment]
I confirmed that the new Travis failures occurring in this branch were from upstream changes. Closed.

**Pull request merged**